ENH added code to rotate anaconda tokens

beckermr opened the pull request:
This PR is cleaned up version of the token rotation code from #1283. I will rebase that PR on master once this one goes in.

<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Added a ``news`` entry

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

beckermr commented:
OK @conda-forge/core @isuruf @scopatz This one is ready to be reviewed and merged!

beckermr commented:
@conda-forge/core can someone check this over? It is needed for CFEP-13.

isuruf commented:
Can we use another name? Otherwise there'll be disruption between the time the token is changed and there's a rerender.

beckermr commented:
Another name for what? The token on disk?

isuruf commented:
Name of the token. (`BINSTAR_TOKEN`)

beckermr commented:
I'm confused about the issue here. Can you be more specific? Using this command does not require a rerender except for appveyor.

isuruf commented:
If I'm understanding this correctly, this PR is used for changing the BINSTAR_TOKEN for `conda-forge` account into `cf-staging`. AFAIK, to use `cf-staging`, you need a rerender, right?

beckermr commented:
Ahhhh so I think we might be conflating two different things here. On the one hand, we could use the code here to rotate the binstar tokens we have. This utility seems quite general. 

On the other, I do plan to use this for cfep-13. My plan was to rerender each feedstock and push straight to master without a rebuild. A full rebuild would be expensive and seems unnecessary. I know this procedure wouldn’t be great for reproducibility, but rebuilding every feedstock in conda forge is a big task and not needed. If we are concerned about reproducibility we could add a git tag to each feedstock before we push so that we can mark the last build before the change. 

Even if we don’t end up with the plan above, the code in this pr is still quite general and so could be merged as is.

beckermr commented:
Btw, if we do end up wanting a separate token name, then we can push a change later to have the token name be configurable. I could add that to this pr now if you’d like too.

isuruf commented:
> My plan was to rerender each feedstock and push straight to master without a rebuild. 

That's too invasive. This would make all PRs stale and some feedstocks have some custom tweaks that they reapply every rerender.

> Btw, if we do end up wanting a separate token name, then we can push a change later to have the token name be configurable. I could add that to this pr now if you’d like too.

Sure.

isuruf commented:
Also, I'd like to have feedstock master working and reproducible. A forced rerender might bring in changes that make the feedstock not work anymore.

beckermr commented:
> That's too invasive. This would make all PRs stale and some feedstocks have some custom tweaks that they reapply every rerender.

I understand the concern about that for sure, but all PRs will be stale eventually anyways. We have to turn off the old token at some point. Further, even if we issued a rebuild PR to every feedstock and a large fraction of them were merged, we'd still leave the current PRs on the feedstocks stale until they were rerendered (if they had merged the CFEP-13 rebuild or not). I don't see the actual difference between the procedure I suggested above and the alternative and making true rebuild PRs with separate tokens.

beckermr commented:
The downside of full rebuild PRs is that they are quite taxing on our volunteer maintainers.

beckermr commented:
> Also, I'd like to have feedstock master working and reproducible. A forced rerender might bring in changes that make the feedstock not work anymore.

I agree. We can apply a git tag so we can recover the feedstock as it was.

isuruf commented:
My suggestion was not to make full rebuild PRs. Sorry for not making it clear.
My suggestion is to add both `BINSTAR_TOKEN` and `STAGING_BINSTAR_TOKEN` to `conda-forge.yml` and when there's a new rerender by a user with the new conda-smithy, it will rerender with `STAGING_BINSTAR_TOKEN`. After a few weeks then we remove `BINSTAR_TOKEN`.

beckermr commented:
> My suggestion was not to make full rebuild PRs. Sorry for not making it clear.
My suggestion is to add both BINSTAR_TOKEN and STAGING_BINSTAR_TOKEN to conda-forge.yml and when there's a new rerender by a user with the new conda-smithy, it will rerender with STAGING_BINSTAR_TOKEN. After a few weeks then we remove BINSTAR_TOKEN.

Sure we can do that, but most feedstocks don't see regular rebuilds. So even if we have an in-between state, we eventually leave a bunch of PRs and feedstocks behind.

isuruf commented:
True. My main concern is for PRs already with a green button which were rerendered with an old version. If we force a rerender on master (which is not good for reproducability), these would become stale. If we don't, after merging these would fail. That's why I suggested a compromise.

beckermr commented:
Sounds good. I’ll need to push an adjustment to the ci setup package to use the staging token if validating.

beckermr commented on conda_smithy/feedstock_tokens.py:
```diff
@@ -618,6 +618,8 @@ def add_feedstock_token_to_azure(user, project, feedstock_token, clobber):
             "Cannot add FEEDSTOCK_TOKEN to a repo that is not already registerd on azure CI!"
         )
 
+    ed = bclient.get_definition(ed.id, project=config.project_name)
```
This was a bug I found.

beckermr commented:
OK @isuruf. I put in the requested changes. I fixed three bugs.

1. The code now checks the conda-forge.yml to make sure any travis tokens there with the same name are gone. These tokens will override the ones set in the API. It also warns the the feedstock will need to be rerendered in certain cases when updating tokens in travis.
2. I fixed a bug in the feedstock token registration that deleted other secrets on azure.
3. I fixed the argparse calls so that the output of the subcommands is correct.

This one is ready for another look!

beckermr commented:
thanks!
